ddtrace/llmobs/_integrations/utils.py

```diff
@@ -397,6 +397,209 @@ def openai_set_meta_tags_from_chat(span: Span, kwargs: Dict[str, Any], messages:
+                        "trace_id": format_trace_id(span.trace_id),
+                        "span_id": str(span.span_id),
+                    },
+                ),
+            )
+            if tool_call_info:
+                output_messages.append({"content": "", "tool_calls": tool_call_info})
+                core.dispatch(DISPATCH_ON_TOOL_CALL_OUTPUT_USED, (tool_id, span))
+        span._set_ctx_item(OUTPUT_MESSAGES, output_messages)
+
+
+def openai_construct_response_from_streamed_chunks(streamed_chunks: List[Any]) -> Dict[str, Any]:
+    """Constructs a response dictionary from streamed chunks.
+    The resulting response dictionary is of form:
+    {"output": [{"role": "...", "content": [{"text": "..."}]}]}
+    """
+    message: Dict[str, Any] = {"content": "", "tool_calls": []}
+
+    for chunk in streamed_chunks:
+        if not isinstance(chunk, tuple) or len(chunk) != 2:
+            continue
+
+        chunk_type, chunk_value = chunk
+        # print(f"Processing chunk type: {chunk_type}, value: {chunk_value}")
+        if chunk_type == "usage":
+            """
+            TODO: currently chunk_value is a ResponseUsage object.
+            We want to convert it to a dict that looks like 
+            {"input_tokens": ..., "output_tokens": ..., "total_tokens": ...}
+            """
+            message["usage"] = chunk_value
+
+        if chunk_type == "output" and isinstance(chunk_value, list):
+            for output in chunk_value:
+                if hasattr(output, "content"):
+                    # content is a list of ResponseOutputText objects
+                    for content in output.content:
+                        if hasattr(content, "text"):
+                            message["content"] += content.text
+                    output_role = output.role if hasattr(output, "role") else ""
+                    message["role"] = output_role
+                if hasattr(output, "summary"):
+                    for summary in output.summary:
+                        # summary_text = summary.text
+                        message["content"] += summary.text
+                        message["role"] = "reasoning"
+
+                elif hasattr(output, "type"):
+                    if output.type in ["function_call", "file_search_call", "web_search_call", "computer_call"]:
+                        tool_name = output.name if hasattr(output, "name") else ""
+                        tool_type = output.type if hasattr(output, "type") else ""
+                        tool_id = output.id if hasattr(output, "id") else ""
+                        tool_arguments = output.arguments if hasattr(output, "arguments") else {}
+                        # file search
+                        tool_queries_raw = output.queries if hasattr(output, "queries") else []
+                        tool_query = ""
+                        if tool_queries_raw:
+                            tool_query = json.dumps({"queries": tool_queries_raw})
```

Comment:
Why are we json dumping and then json loading later? json encoding/decoding is expensive 😢